File: typescript/src/monsters/types/undead.ts
import { parseActiveAbility, parseAttack, passiveAbilities } from "@src/monsters/mechanics";
import { MonsterBaseInput } from "@src/monsters/reformat_monster_input";
import { addType, TypelessMonsterInput } from "./add_type";

export const undeadInput: TypelessMonsterInput[] = [
  {
    alignment: "Always neutral evil",
    attackInputs: [
      {
        crit: "The target loses two \\glossterm{hit points}.",
        defense: "reflex",
        hit: "The target loses a \\glossterm{hit point}.",
        source: "magical",
        name: "Draining Touch",
        target: "One living creature within \\glossterm{reach}",
      },
    ],
    challengeRating: 4,
    knowledge: {
      0: `
        An allip is a Medium incorporeal creature.
        It cannot speak intelligibly, though it is known for its propensity for babbling incoherently as it attacks.
      `,
      5: `
        An allip is spectral remains of someone driven to suicide by a madness that afflicted it in life.
        It craves only revenge and unrelentingly pursues those who tormented it in life and pushed it over the brink.
      `,
    },
    level: 3,
    name: "allip",
    passiveAbilities: [
      passiveAbilities.incorporeal,
      {
        description: `
          During each \\glossterm{action phase}, the allip makes an attack vs. Mental against each creature
          within an \\areamed radius \\glossterm{emanation} from it.
          After it attacks a creature this way, that creature becomes immune to this allip's \\textit{babble} ability until it takes a \\glossterm{short rest}.
          \\hit Each target is \\dazed as a \\glossterm{condition}.
          \\crit Each target is \\confused as a \\glossterm{condition}.
        `,
        name: "Babble",
      },
    ],
    skillPoints: { awareness: 2, intimidate: 2, stealth: 2 },
    speeds: { fly: 30, land: null },
    startingAttributes: { str: null, dex: 3, con: null, int: 1, per: 1, wil: 2 },
  },
];

function skelefy(monster: Omit<MonsterBaseInput, "monsterType">): MonsterBaseInput {
  return {
    ...monster,
    activeAbilityInputs: (monster?.activeAbilityInputs || []).filter((ability) => {
      const parsed = parseActiveAbility(ability);
      return parsed.source === "mundane";
    }),
    alignment: "Always true neutral",
    attackInputs: (monster?.attackInputs || []).filter((attack) => {
      const parsed = parseAttack(attack);
      return parsed.source === "mundane" && !parsed.weaponName;
    }),
    // TODO: Should description be cleared?
    knowledge: {},
    monsterType: "undead",
    name: `Skeleton ${monster.name}`,
    startingAttributes: {
      ...monster.startingAttributes,
      str: (monster.startingAttributes?.str || 0) - 2,
      dex: (monster.startingAttributes?.dex || 0) + 2,
      int: null,
    },
    passiveAbilities: [
      // TODO: filter magical passive abilities
      ...(monster.passiveAbilities || []),
      {
        description: `
          Skeletons are \\glossterm{vulnerable} to bludgeoning damage.
        `,
        name: "Bludgeoning Vulnerability",
      },
    ],
    weaponInput: [
      ...(monster.weaponInput || []).filter((weapon) => weapon.name !== "claw"),
      { name: "claw" },
    ],
  };
}

function zombify(monster: Omit<MonsterBaseInput, "monsterType">): MonsterBaseInput {
  return {
    ...monster,
    activeAbilityInputs: (monster?.activeAbilityInputs || []).filter((ability) => {
      const parsed = parseActiveAbility(ability);
      return parsed.source === "mundane";
    }),
    alignment: "Always true neutral",
    attackInputs: (monster?.attackInputs || []).filter((attack) => {
      const parsed = parseAttack(attack);
      return parsed.source === "mundane" && !parsed.weaponName;
    }),
    // TODO: Should description be cleared?
    knowledge: {},
    monsterType: "undead",
    name: `Zombie ${monster.name}`,
    startingAttributes: {
      ...monster.startingAttributes,
      str: (monster.startingAttributes?.str || 0) + 2,
      dex: (monster.startingAttributes?.dex || 0) - 2,
      con: (monster.startingAttributes?.con || 0) + 2,
      int: null,
    },
    passiveAbilities: [
      // TODO: filter magical passive abilities
      ...(monster.passiveAbilities || []),
      {
        description: `
          Zombies are \\glossterm{vulnerable} to slashing damage.
        `,
        name: "Slashing Vulnerability",
      },
    ],
    weaponInput: [{ name: "slam" }],
  };
}

// const orcs = humanoids.find((mon) => mon.name === "Orcs") as MonsterGroupInput;
// const zombifiableMonsters = orcs.monsters.filter((orc) => orc.name !== "Orc Shaman");
const zombifiableMonsters: any[] = [];

undeadInput.push({
  knowledge: {
    0: `
      Zombies are the reanimated corpses of once-living creatures.
      They are one of the most basic forms of undead, though some zombies are still quite powerful.
      Creating a zombie is generally regarded as a fundamentally evil act.
    `,
    5: `
      Zombies retain all of the \\glossterm{mundane} abilities of the reanimated creature, but lose all \\glossterm{magical} abilities.
      They also lose any ability to wield weapons, though they are often found in the same armor (if any) as the original creature.
      In addition, zombies are always stronger and more durable than the original creature, though they are less agile.
      All zombies are \\glossterm{vulnerable} to slashing damage thanks to their exposed and rotting flesh.
    `,
    10: `
      Creating a zombie from a corpse requires fracturing the soul of the creature the corpse belonged to.
      The soul splinter created this way is used to give the zombie its agency.
      This is unpleasant for the dead creature in its afterlife, though not dangerous.
      Only corpses with the majority of their flesh intact can be reanimated as zombies.
      Creatures with too little flesh cannot become zombies at all.
    `,
  },
  level: 1,
  name: "Zombies",
  monsters: zombifiableMonsters.map(zombify),
});

export const undead = addType("undead", undeadInput);

File: rust/src/monsters/specific_monsters/undead.rs
use crate::core_mechanics::{
    Attribute, DamageType, HasAttributes, MovementMode, Sense, Size, SpecialDefenseType,
    StandardPassiveAbility,
};
use crate::creatures::{Modifier, Monster};
use crate::equipment::{StandardWeapon, Weapon};
use crate::monsters::challenge_rating::ChallengeRating;
use crate::monsters::creature_type::CreatureType::Undead;
use crate::monsters::knowledge::Knowledge;
use crate::monsters::monster_entry::MonsterEntry;
use crate::monsters::monster_group::MonsterGroup;
use crate::monsters::{monster_group, FullMonsterDefinition};
use crate::skills::Skill;
use std::cmp::{max, min};

use super::humanoids::add_humans;

struct FullUndeadDefinition {
    alignment: String,
    attributes: Vec<i32>,
    challenge_rating: ChallengeRating,
    description: Option<String>,
    knowledge: Option<Knowledge>,
    level: i32,
    modifiers: Option<Vec<Modifier>>,
    movement_modes: Option<Vec<MovementMode>>,
    name: String,
    senses: Option<Vec<Sense>>,
    size: Size,
    trained_skills: Option<Vec<Skill>>,
    weapons: Vec<Weapon>,
}

impl FullUndeadDefinition {
    fn monster(self) -> Monster {
        let mut modifiers = self.modifiers.unwrap_or(vec![]).clone();
        modifiers.push(Modifier::PassiveAbility(
            StandardPassiveAbility::Undead.ability(),
        ));
        return FullMonsterDefinition {
            // From self
            alignment: self.alignment,
            attributes: self.attributes,
            challenge_rating: self.challenge_rating,
            description: self.description,
            knowledge: self.knowledge,
            level: self.level,
            modifiers: Some(modifiers),
            movement_modes: self.movement_modes,
            name: self.name,
            senses: self.senses,
            size: self.size,
            trained_skills: self.trained_skills,
            weapons: self.weapons,

            // Default values
            creature_type: Undead,
        }
        .monster();
    }
}

pub fn undeads() -> Vec<MonsterEntry> {
    let mut monsters: Vec<MonsterEntry> = vec![];

    add_skeletons(&mut monsters);
    add_zombies(&mut monsters);

    let mindless = Modifier::PassiveAbility(StandardPassiveAbility::Mindless.ability());

    let zombie_vulnerability =
        Modifier::Vulnerable(SpecialDefenseType::Damage(DamageType::Slashing));

    monsters.push(MonsterEntry::MonsterGroup(monster_group::MonsterGroup {
        name: "Zombies".to_string(),
        knowledge: None,
        monsters: vec![
            FullUndeadDefinition {
                alignment: "Always neutral evil".to_string(),
                attributes: vec![3, -2, 3, 0, 0, -2],
                challenge_rating: ChallengeRating::Half,
                description: None,
                knowledge: None,
                level: 1,
                modifiers: Some(vec![mindless.clone(), zombie_vulnerability.clone()]),
                movement_modes: None,
                name: "Zombie Shambler".to_string(),
                senses: None,
                size: Size::Medium,
                trained_skills: None,
                weapons: vec![StandardWeapon::Slam.weapon()],
            }
            .monster(),
            FullUndeadDefinition {
                alignment: "Always neutral evil".to_string(),
                attributes: vec![4, -2, 4, 0, 0, -2],
                challenge_rating: ChallengeRating::One,
                description: None,
                knowledge: None,
                level: 1,
                modifiers: Some(vec![mindless.clone(), zombie_vulnerability.clone()]),
                movement_modes: None,
                name: "Zombie Walker".to_string(),
                senses: None,
                size: Size::Medium,
                trained_skills: None,
                weapons: vec![StandardWeapon::Slam.weapon()],
            }
            .monster(),
            FullUndeadDefinition {
                alignment: "Always neutral evil".to_string(),
                attributes: vec![5, -2, 5, 0, 0, -2],
                challenge_rating: ChallengeRating::Two,
                description: None,
                knowledge: None,
                level: 2,
                modifiers: Some(vec![mindless.clone(), zombie_vulnerability.clone()]),
                movement_modes: None,
                name: "Zombie Brute".to_string(),
                senses: None,
                size: Size::Medium,
                trained_skills: None,
                weapons: vec![StandardWeapon::Slam.weapon()],
            }
            .monster(),
            FullUndeadDefinition {
                alignment: "Always neutral evil".to_string(),
                attributes: vec![6, -2, 6, 0, 0, -2],
                challenge_rating: ChallengeRating::Four,
                description: None,
                knowledge: None,
                level: 3,
                modifiers: Some(vec![mindless.clone(), zombie_vulnerability.clone()]),
                movement_modes: None,
                name: "Zombie Hulk".to_string(),
                senses: None,
                size: Size::Large,
                trained_skills: None,
                weapons: vec![StandardWeapon::Slam.weapon()],
            }
            .monster(),
        ],
    }));

    return monsters;
}

fn add_skeletons(monsters: &mut Vec<MonsterEntry>) {
    let mut corpses = vec![];
    let mut skeletons = vec![];
    add_humans(&mut corpses);

    for entry in corpses {
        if let MonsterEntry::MonsterGroup(group) = entry {
            for monster in group.monsters {
                skeletons.push(convert_to_skeleton(monster));
            }
        } else if let MonsterEntry::Monster(monster) = entry {
            skeletons.push(convert_to_skeleton(monster));
        }
    }

    monsters.push(MonsterEntry::MonsterGroup(MonsterGroup {
        name: "Skeletons".to_string(),
        knowledge: Some(Knowledge::new(vec![
            (0, "
                Skeletons are the reanimated corpses of once-living creatures.
                They are the most basic form of animated undead, since they can be created from corpses that have been reduced to mere bones.
                Creating a skeleton is generally regarded as a fundamentally evil act.
            "),
            (5, "
                Skeletons retain all of the \\glossterm{mundane} abilities of the reanimated creature, but lose all \\glossterm{magical} abilities.
                They retain the ability to wield the same weapons and armor as the original creature, but they are completely mindless.
                In addition, skeletons are always more agile and less strong than the original creature.
                All skeletons are vulnerable to bludgeoning damage thanks to their exposed and easily crumpled bones.
            "),
            (10, "
                Creating a skeleton from a corpse requires splintering the soul of the creature the corpse belonged to.
                The soul splinter created this way is used to give the skeleton its agency.
                This is unpleasant for the dead creature in its afterlife, though not dangerous.

                Skeletons are never created by ambient necromantic magic.
                They have no internal intelligence or agency of any kind, and precisely obey the instructions of their controllers.
                If their instructions are poorly worded or incomplete, skeletons may fail to fight even if attacked.
            "),
        ])),
        monsters: skeletons,
    }));
}

fn convert_to_skeleton(monster: Monster) -> Monster {
    let creature = monster.creature;
    // +1 str, +1 dex, -2 con, fixed int/per/wil
    let max_attribute = monster.challenge_rating.max_base_attribute();
    let attributes = vec![
        min(
            max_attribute,
            creature.get_base_attribute(&Attribute::Strength) + 1,
        ),
        min(
            max_attribute,
            creature.get_base_attribute(&Attribute::Dexterity) + 1,
        ),
        max(
            -9,
            creature.get_base_attribute(&Attribute::Constitution) - 2,
        ),
        -7,
        0,
        0,
    ];

    let mut modifiers = vec![
        Modifier::Vulnerable(SpecialDefenseType::Damage(DamageType::Bludgeoning)),
    ];
    for im in creature.identified_modifiers {
        if im.source == "FullMonsterDefinition" && !im.modifier.is_magical() {
            modifiers.push(im.modifier.clone());
        }
    }
    modifiers.push(Modifier::PassiveAbility(StandardPassiveAbility::Mindless.ability()));

    let mut senses = creature.senses.unwrap_or(vec![]).clone();
    if !senses.iter().any(|s| {
        if let Sense::Darkvision(_) = s {
            true
        } else {
            false
        }
    }) {
        senses.push(Sense::Darkvision(60));
    }

    return FullUndeadDefinition {
        alignment: "Always neutral evil".to_string(),
        attributes,
        challenge_rating: monster.challenge_rating,
        description: monster.description,
        knowledge: monster.knowledge,
        level: creature.level,
        modifiers: Some(modifiers),
        movement_modes: Some(creature.movement_modes),
        name: format!("Skeletal {}", creature.name.unwrap()),
        senses: Some(senses),
        size: creature.size,
        trained_skills: None,
        weapons: creature.weapons,
    }
    .monster();
}

fn convert_to_zombie(monster: Monster) -> Monster {
    let creature = monster.creature;
    // +2 str, -2 dex, +2 con, fixed int/per/wil
    let max_attribute = monster.challenge_rating.max_base_attribute();
    let attributes = vec![
        min(
            max_attribute,
            creature.get_base_attribute(&Attribute::Strength) + 2,
        ),
        max(
            -9,
            creature.get_base_attribute(&Attribute::Dexterity) - 2,
        ),
        min(
            max_attribute,
            creature.get_base_attribute(&Attribute::Constitution) + 2,
        ),
        -7,
        0,
        -3,
    ];

    let mut modifiers = vec![
        Modifier::Vulnerable(SpecialDefenseType::Damage(DamageType::Slashing)),
    ];
    for im in creature.identified_modifiers {
        if im.source == "FullMonsterDefinition" && !im.modifier.is_magical() {
            modifiers.push(im.modifier.clone());
        }
    }

    let mut senses = creature.senses.unwrap_or(vec![]).clone();
    if !senses.iter().any(|s| {
        if let Sense::Darkvision(_) = s {
            true
        } else {
            false
        }
    }) {
        senses.push(Sense::Darkvision(60));
    }

    return FullUndeadDefinition {
        alignment: "Always neutral evil".to_string(),
        attributes,
        challenge_rating: monster.challenge_rating,
        description: monster.description,
        knowledge: monster.knowledge,
        level: creature.level,
        modifiers: Some(modifiers),
        movement_modes: Some(creature.movement_modes),
        name: format!("Zombie {}", creature.name.unwrap()),
        senses: Some(senses),
        size: creature.size,
        trained_skills: None,
        weapons: creature.weapons,
    }
    .monster();
}


fn add_zombies(monsters: &mut Vec<MonsterEntry>) {
    let mut corpses = vec![];
    let mut zombies = vec![];
    add_humans(&mut corpses);

    for entry in corpses {
        if let MonsterEntry::MonsterGroup(group) = entry {
            for monster in group.monsters {
                zombies.push(convert_to_zombie(monster));
            }
        } else if let MonsterEntry::Monster(monster) = entry {
            zombies.push(convert_to_zombie(monster));
        }
    }

    monsters.push(MonsterEntry::MonsterGroup(MonsterGroup {
        name: "Zombies".to_string(),
        knowledge: Some(Knowledge::new(vec![
            (0, "
                Zombies are the reanimated corpses of once-living creatures.
                They must be created from corpses that still retain most of their organs and internal structure.
                Creating a zombie is generally regarded as a fundamentally evil act.
            "),
            (5, "
                Zombies retain all of the \\glossterm{mundane} abilities of the reanimated creature, but lose all \\glossterm{magical} abilities.
                They retain the ability to wield the same weapons and armor as the original creature.
                In addition, zombies are always stronger and less agile than the original creature.
                All zombies are vulnerable to slashing damage thanks to their exposed and easily torn skin and muscles.
            "),
            (10, "
                Creating a zombie from a corpse requires splintering the soul of the creature the corpse belonged to.
                The soul splinter created this way is used to give the zombie its agency.
                This is unpleasant for the dead creature in its afterlife, though not dangerous.

                Zombies are sometimes created by ambient necromantic magic.
                Even if they are created and controlled by necromancers, they still retain an animalistic hunger for flesh, especially brains.
                If their instructions are poorly worded or incomplete, zombies may attack any living creature they see.
            "),
        ])),
        monsters: zombies,
    }));
}
